.github/workflows/build-wheels.yaml

```
@@ -0,0 +1,116 @@
+name: build-wheels
+run-name: build-wheels
+
+on:
+  push:
+    # only build+push docker on release-please tags
+    tags: [ "v*" ]
+    paths:
+    - ".github/actions/**"
+    - ".github/workflows/**"
+    - "conda-recipe/**"
+    - "genome_kit/**"
+    - "setup.py"
+    - "setup/**"
+    - "src/**"
+    - "tests/**"
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+        # optional but recommended for speed.
+        # the docker image in the workflow is necessary to build a manylinux compliant wheel.
+        # PyPI expects these to be run in manylinux_2_28 or lower.
+        # additionally, manylinux_2_28 is actively maintained by PyPA and includes /opt/python/ environments
+        # for 3.9-3.12 out-of-the-box.
+      - name: Pre-pull manylinux image
+        run: docker pull quay.io/pypa/manylinux_2_28_x86_64
+
+      - name: Build and repair wheels for Python 3.9–3.12
+        run: |
+          docker run --rm -v $(pwd):/io -e BUILD_WHEELS=1 quay.io/pypa/manylinux_2_28_x86_64 /bin/bash -c "
+            set -euxo pipefail
+
+            PYTHONS=(
+              /opt/python/cp39-cp39/bin/python
+              /opt/python/cp310-cp310/bin/python
+              /opt/python/cp311-cp311/bin/python
+              /opt/python/cp312-cp312/bin/python
+            )
+
+            # Build and install libfmt to statically link to wheel
+            cd /tmp
+            # for now 11.1.4 is broken, otherwise use latest tag
+            git clone --branch 11.0.2 --depth=1 https://github.com/fmtlib/fmt.git
+            cd fmt
+            cmake -DCMAKE_POSITION_INDEPENDENT_CODE=TRUE -DFMT_INSTALL=ON .
+            make -j\$(nproc)
+            make install
+            cd /io
+
+            mkdir -p repaired_wheels
+
+            # install build dependencies
+            # numpy is required at build time because its c api is used
+            for PYTHON in \${PYTHONS[@]}; do
+              \$PYTHON -m pip install -U pip setuptools wheel cmake auditwheel numpy
```

Match [meta.yaml](https://github.com/conda-forge/genomekit-feedstock/blob/e14f9f0874649f8cf36fa53262f2755e937f0c3a/recipe/meta.yaml#L22)

```suggestion
              \$PYTHON -m pip install -U pip setuptools wheel cmake auditwheel numpy<2
```

---

I don't suggest this since it will make numpy=2 incompatible, see the above comment thread

---

agreed. I will try @ovesh's suggestion to use `setup_requires`. I'll have to test if these properties get inherited into the Docker image in which the wheel is built:

> Wouldn't this work with `setup_requires` in setup.py? (at least for wheel, cmake, auditwheel, and numpy).

The alternative to this would be to define the version inline as `numpy<2.0dev0` or `numpy>=1.25,<2.0dev0` to set a lower bound.

---

it turns out `setup_requires` field in `setup.py` is insufficient in this case due to the following:

- although `setup_requires` is used to specify build-time dependencies need to run `setup.py`, it doesn't install `auditwheel`, `cmake`, nor `wheel` are build _tools_ not _dependencies_ so `setup.py` might have a hard time installing them. 
- more importantly, as I suspected these dependencies would not be inherited by the Docker image. Hence a `pip install` is required up front in the Docker env. 
- Further, `setup_requires` is [deprecated in `setuptools`](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#setup-requires), which is used in the workflow. 

As such I'll apply the alternative I mentioned.

---

have you looked into https://github.com/pypa/cibuildwheel or how it does it?  seems like most compbio projects use this

---

Yes, much of what I tried initially was with `cibuildwheel`, however I encountered numerous difficulties using it. Eventually it was just easier to use `bdist_wheel` instead and manually build for supported platforms and Python versions. I haven't been able to pinpoint exactly what the issues were, but here's some things that I tried. 

`cibuildwheel` runs in an isolated environment just like the Docker image of the current implementation,  the difference being that what happens in the environment is out of our control and cannot be directly edited. 

- GK relies on native library dependencies, namely `libfmt`. These can be installed prior to the build with the [`CIBW_BEFORE_ALL`](https://cibuildwheel.pypa.io/en/stable/options/#before-all) environment variable, as outlined [here](https://cibuildwheel.pypa.io/en/stable/faq/#missing-dependencies). For a specific version of `libfmt`, we can do something like
```
CIBW_BEFORE_ALL_LINUX: |
    curl -L https://github.com/fmtlib/fmt/archive/refs/tags/11.0.2.tar.gz | tar xz
    cd fmt-11.0.2
    cmake -DCMAKE_POSITION_INDEPENDENT_CODE=ON -DBUILD_SHARED_LIBS=OFF -DCMAKE_INSTALL_PREFIX=/usr/local .
    make -j$(nproc)
    make install
```
- GK also relies on build-time dependencies like `cmake` and `numpy`. These can be specified with the [`CIBW_BEFORE_BUILD`](https://cibuildwheel.pypa.io/en/stable/options/#before-build) environment variable in the workflow as 
```
CIBW_BEFORE_BUILD: pip install cmake "numpy>=1.25,<2.0.0.dev0"
```
Running the workflow with the above still results in a `ModuleNotFoundError` for `numpy` when the wheel is built, despite having installed it in a previous step:
```
Running before_build...
...
   Installing collected packages: numpy, cmake
   Successfully installed cmake-4.0.0 numpy-1.26.4

Building wheel...
...
   File "/project/setup/__init__.py", line 3, in <module>
      from .c_ext import *
   File "/project/setup/c_ext.py", line 37, in <module>
      import numpy as np
   ModuleNotFoundError: No module named 'numpy'
```
- Notably, `cibuildwheel` recommends an alternative to these environment variables being the `pyproject.toml`, where build-time dependencies can be specified under `[build-system] requires=[]`. Read more [here](https://cibuildwheel.pypa.io/en/stable/options/#before-build). This results in a new error:
```
   File "/tmp/pip-build-env-la3q964b/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 320, in run_setup
      exec(code, locals())
   File "<string>", line 6, in <module>
   ModuleNotFoundError: No module named 'setup'
```
To attempt a fix I tried deliberately binding the path with `os.path.join(os.path.dirname(__file__), "setup")` which is considered bad practice, which also didn't work. 
I also tried renaming the `/setup` dir to avoid ambiguity between `/setup` and `setup.py` for `from setup import c_ext`, which also resulted in the same `ModuleNotFoundError` as before. 

Furthermore, I was concerned that even if I could get `cibuildwheel` to work, given that parts of GK rely on the `C++20` standard, we'd have trouble using `cibuildwheel` as their [docs](https://cibuildwheel.pypa.io/en/stable/cpp_standards/) only reference support of up to `C++17`. 
Regardless `cibuildwheel` for linux by default builds for `manylinux_2014_x86` which does not support `C++20` standard. Configuring for a more modern `manylinux` image like `manylinux_2_28_x86` as used in the current implementation requires fidgeting with more environment variable options which I didn't get to researching. The thread below discusses compatibility with more modern `C++` standards.

(https://github.com/pypa/cibuildwheel/discussions/984)

As you've stated @s22chan many projects use `cibuildwheel` but there seem to be some obstacles with it that I was unable to overcome, hence the implementation with `bdist_wheel`.

---

thanks, that's a great comprehensive paper trail on the alternatives tried

---

Unresolving this since the original issue of numpy=2 incompatibility was not resolved

---

@s22chan I followed the version specified in `genomekit_dev.yml`:
```
dependencies:
  - appdirs>=1.4.0
  - fmt
  - numpy <2.0dev0
```
I'm a little confused, did you mean something like `"numpy>=1.25,=<2.0.0.dev0"` rather than the current `"numpy>=1.25,<2.0.0.dev0"`?

---

I might have read this wrong, but I think it's building with numpy<2.

from https://numpy.org/doc/stable/dev/depending_on_numpy.html#numpy-2-0-specific-advice:

> NumPy 2.0 is an ABI-breaking release, however it does contain support for building wheels that work on both 2.0 and 1.xx releases. It’s important to understand that:
>
>    When you build wheels for your package using a NumPy 1.xx version at build time, those will not work with NumPy 2.0.
>
>    When you build wheels for your package using a NumPy 2.x version at build time, those will work with NumPy 1.xx.

---

this makes sense. I'll change it to `<=2` for now.

---

note to self: add unit test runs explicitly for numpy v1 and numpy v2

---

I'm going to go ahead and merge, since there's nothing being published yet and the conda package is unaffected. You can add the tests later.